### PR TITLE
Specifying Julia executable/args via RuntimeEnv

### PR DESCRIPTION
**Pull request by omus**
Depends on: https://github.com/beacon-biosignals/ray/pull/8

Closes: #103, #100

TODO:
- [x] Add tests

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205478658766583

### CODE REVIEW

**Comment by omus** [comment]
I will validate that #100 is actually addressed locally tomorrow. I did do a test to ensure that `--code-coverage` is being passed through but haven't done a full end-to-end test.

**Comment by omus** [comment]
Things for tomorrow:
- Allow user specified Julia executable? Remote nodes may not have Julia installed at the same location
- BASH injection vulnerability with Ray's runtime env code?
- Specify additional default flags? e.g. `--startup-file=no`
- Combine `executable` and `args`?
- Deprecate `py_executable`
- Test code coverage from drivers and workers don't overwrite coverage files

**Comment by omus** [comment]
> Allow user specified Julia executable? Remote nodes may not have Julia installed at the same location

[Distributed.jl does allow end users to specify the path to the Julia executable](https://github.com/JuliaLang/julia/blob/87e31a5d5e054dd2e48fab0c8ce4f7e8f7facce7/stdlib/Distributed/src/cluster.jl#L23-L24). In Ray.jl specifying the Julia executable path would only really allow for having the driver and worker Julia paths to differ but all workers would need to be the same. Possibly specifying a Ray task with resource requirements could allow further heterogeneity. 

> BASH injection vulnerability with Ray's runtime env code

In evaluating the original ray-2.5.1 code the only two parameters that are user definable are `env_vars` and `java_jars`. All other variables can not be specified via a passed in runtime environment. Since `env_vars` only set values in `os.environ` this cannot result in arbitrary code execution. It may be possible using `java_jars` to inject a carefully executed bash script which contained no spaces (e.g. `{touch,file}`) as the only escaping of user specified `java_jars` is handled by replacing spaces with escaped spaces. I have not checked the Java front end but if they validate the contents of `java_jars` before hand the threat can be mitigated.

Our update to `context.py` where we allow the client to specify how worker processes are spawned may be problematic if the Ray devs are highly concerned about this problem. Without this kind of support we may be limited to only launching the the Julia process on the path named `julia`.

> Specify additional default flags? e.g. `--startup-file=no`

Any flags that `julia_cmd()` passes through will be passed in automatically. We can always add additional flags in later if we want them.

> Combine `executable` and `args`?

I decided this was a good idea.

> Deprecate `py_executable`

@kleinschmidt [brought up](https://github.com/beacon-biosignals/ray/pull/8#discussion_r1326047775) that the same `RuntimeEnvContext` may be used to launch workers using different languages. Because of this we should leave `py_executable` as is 

> Test code coverage from drivers and workers don't overwrite coverage files

It works! Each coverage file includes the PID in the file name. This means our workers and driver coverage files aren't clobbering each other.

> Client side tests

Done

**Review by kleinschmidt — APPROVED** [review]

File: src/runtime_env.jl [hunk 1]
```diff
@@ -22,9 +22,12 @@ function json_dict(runtime_env::RuntimeEnv)
         env_vars["JULIA_RAY_PACKAGE_IMPORTS"] = base64encode(serialize, imports)
     end
 
+    code = "using $(@__MODULE__); start_worker()"
+    cmd = `$(Base.julia_cmd()) -e $code`
+
     # The keys of `context` must match what is supported by the Python `RuntimeEnvContext`:
     # https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/_private/runtime_env/context.py#L20-L45
-    context = Dict("env_vars" => env_vars)
+    context = Dict("julia_command" => cmd.exec, "env_vars" => env_vars)
```
[the remark this answers is not in the document]
Comment (omus) [reply src/runtime_env.jl:31]: Your understanding is correct

**Pull request merged**